### ISSUE TITLE
differential_evolution bug converges to wrong results in complex cases

### ISSUE THREAD

**Issue by yonadav** [scipy/scipy#8256]
The differential evolution does not converge correctly, and is incompatible with the original code (from algorithmic point of view) of Storn.
Storn claims that the algorithm has negligible advantage when increasing the population over 40 regardless of the number of objective function model variables. When I tried to use a population of ~60 with my problem which is quite complex (Solving Arrhenius parameters for several reactions). It converged to wrong results and the objective function was far from the correct values. While checking the code I found that the population which is used during an iteration is written during the same iteration. It creates an unacceptable bias in the direction of the currently best  result and all the population quickly converges to the same result which is wrong. In all the codes I found on the net including Storn`s code the population  used within an iteration is from previous iteration only, and the currently best result is saved in a different array designated for that purpose. at the end of the iteration the modified population is copied to the population to be used at the next iteration.
Privately fixing the code not to modify the population used in the same iteration solved the problem and the function converged correctly with the same small population. Larger populations partially "hide" the bug since it is a stochastic process and the bias becomes minor.
best1bin strategy was used. actually most o the arguments were the default ones, only maxiter, tol, callback, and disp were modified (larger maxiter and smaller tol)
I use Scipy version 0.18.1 from Anaconda distribution.

Another minor (not a bug but structural, implementation problem) problem. An attempt to use the strategy functions as generic preassigned in the constructor was made (very good), alas, not all functions had the same interface which results the need for if statement in the code, the function is not generic any more.
Attached is a fixed version of _differential_evolution.py with the bug and the second problems fixed, as I figured out. Not necessarily with the naming convention of the library.

[_differentialevolution.zip](https://github.com/scipy/scipy/files/1595085/_differentialevolution.zip)

**Comment by yonadav** [comment]
There was actually another severe bug fixed. In the original class when the temporary tested  candidate created by _mutate the is better than the original one it is overridden (O.K). The problem is that if it was better than the currently best, than the 0 index population member was overridden rather than switched with the new best result (in the case where the tested candidate was not the first one in the population, but still was better than the one in the zero index). The previously best is lost as a population member and 2 copies of the new best now exist.

**Comment by ilayn** [comment]
@yonadav Do you mind opening a pull request for this? It is not very convenient to download and setup everything from scratch let alone being able to use GitHub's review systems and so forth. 

Also if there are correctness/performance claims it is best to show the timings and results with a test suite.

**Comment by yonadav** [comment]
There is no performance issue it is simply a bug, since it is a statistical process the bug may be obscured in many cases but it is still there. Verified against the C++ code published by Storn (one of the creators of the algorithm). I cannot publish the code for the objective function and the data-set of the experiment.
My code is just an example of a possible option to solve the problem, did not pay attention to coding conventions of the library. So my guess is that it is not a very good candidate for pull request and was given only as an algorithmic example of the solution to back up the explanation.
The code is an add-on not a must I could just leave the explanations.
If you think it is feasible for a pull request I can open one.
As for tests it should converge for a population of 40 as Storn claimed and it did after the correction for a very complex equation with 24 parameters to solve. It converged successfully for several different data-sets

**Comment by andyfaff** [comment]
The scipy version of `differential_evolution` is the variant outlined in Wormington et al., Phil. Trans. R. Soc. Lond. A (1999) 357, 2827–2848, 10.1098/rsta.1999.0469. The flowchart in fig1 outlines the optimization.
In this variant if the trial vector is better than the parameter vector then it replaces it, if it's better than the best fit vector it also replaces that - *within* the same generation. This leads to faster convergence. You are correct in saying that the code is not the same as the original Storn and Price algorithm, it's simply a variant.
Variants aren't bad if they lead to enhancements. For example, the original Storn and Price code didn't dither the mutation constant, F. However, this has been found to improve convergence behaviour, so it's probably better to have that variant.

**Comment by andyfaff** [comment]
Of course in your situation it seems that you don't want the aggressive variant. It would be relatively straightforward to add an `aggressive=True` as a keyword. If you want to turn off that feature you could supply `aggressive=False`. In that way the current behaviour is unchanged, but you'd also get what you liked.

**Comment by yonadav** [comment]
The reference you have mentioned is not included in the documentation only Wikipedia,  Storn and Price article and Storn`s site. The C++ variant by Storn does include dithering, they do use the currently best result for the strategies which requires the best result within the same generation. However, they use the population from the previous iteration with the exception of the currently best result.
see my second comment, when a new best result it located and it is not on index 0 then both index 0 is overridden as well as the actual candidate location. I find it hard to believe that it is in the algorithm you have mentioned.
It might converge faster in simple cases but in more complex ones it drifts to a totally wrong solution unless you use a very large population. It depends on how computationally expensive  your objective function is. In a very expensive OF having twice the number of iterations but with population of 40 instead of 400 is much better.
As I mentioned before the reference you wrote is not quoted in the manual nor in the code, so I had no way to know about the variant it is based upon. 
I will stick to the variant I made which might converge slower but converges for me. The benefits of open source ....

**Comment by yonadav** [comment]
It would be nice to have that keyword as well as adding the reference to your variant.
But still pay attention to my comment regarding the case when a best solution is located not at 0.
in that case you should switch between the previously best and the current best rather than overwrite it

**Comment by andyfaff** [comment]
I was going to reply to the second comment within the next couple of days.

**Comment by yonadav** [comment]
Hi Andrew
It seems like the major problem was my second comment about the case when the currently new best candidate id is not 0.
I made a simple test based on the function from the help, your original DE does not converge correctly when the population is small enough and the variable range is large enough, Whereas the fixed version (it is a minor fix) converges correctly and with less iterations than the original Storn and Price algorithm.
I attached a zip of the test  _differentialevolutionMod0.py is the version with the correction on the second comment.
I will test it on my considerably more complex problem, where one of the parameters has a naturally possible range of more than 10 orders.

Thanks for adding the extra keyword and for adding the algorithm to scipy it saved me when all other optimization algorithm in the library failed (I was very very close to implement a genetic alg.).

By the way, is it possible to officially expose the  DifferentialEvolutionSolver class for inheritancerather than just the differential_evolution function?

[TestDifferentialEvolution.zip](https://github.com/scipy/scipy/files/1595803/TestDifferentialEvolution.zip)

**Comment by andyfaff** [comment]
I can't read either of the zip files you attached. A gist of the code is a better idea if you want to share things.

> your original DE does not converge correctly when the population is small enough and the variable range is large enough, Whereas the fixed version (it is a minor fix) converges correctly and with less iterations than the original Storn and Price algorithm.

What do you mean by does not converge correctly? Please be precise. Do you mean that it doesn't find a known solution, or does it follow a different path to a solution than the one you expect? In difficult problems (large dimensionality, with low popsize) I wouldn't expect it to find a solution 100% of the time.

Do you agree with the solution I made?

> By the way, is it possible to officially expose the DifferentialEvolutionSolver class for inheritancerather than just the differential_evolution function?

At the moment it's been nice that the class hasn't been exposed, it hides the implementation and has allowed quite large internal changes to be made. That's not to say it can't be made public in the future.

**Comment by yonadav** [comment]
Strange I downloaded the zip from GitHub and opened it, it can be opened in windows integrated compressed folder and with 7Zip. 
What is a gist of the code?
I used the example in the help of differential_evolution (ackley ) but with larger range (-500, 500) and popsize of 7. It converged (from the implementation population convergence tolerance point of view) but to a wrong solution. different wrong solution depends on the seed.
It is a very simple test not structured like the library`s unit testing, just a demonstration of the problem.
It is of no consequences now since you have fixed it. (I saw that you fixed it only after my reply).

Excellent solution (and now I learned how to swap in a single line :) ).

The reason I need the class is because my objective function is highly  computationally expensive and very complex with many variables where some of them have ranges of many orders. It takes several hours  to complete. Hence access to the population and saving the population for a "continue" mode is an advantage (I located _differentialevolution.py and imported _differentialevolution & derived my own class from DifferentialEvolutionSolver class)
Thanks

**Comment by ilayn** [comment]
You can put code in https://gist.github.com/ and can share the link. Then you can make modifications on the code there without the need of attaching files back and forth.

**Comment by yonadav** [comment]
Thanks @ilayn  good to know, I'm new to GitHub.
@andyfaff I downloaded your latest (private) version it now converges correctly on my complex case even with aggressive=True. The bug was probably the need for swapping.

**Comment by andyfaff** [comment]
In these circumstances use of the term 'converge correctly' can be misleading. Correctly has a specific meaning, it implies the code is doing something quite wrong. I'd prefer the term 'doesn't find the global minimum'. Not finding a global minimum is quite common in global optimization, there are no guarantees. Running the benchmark suite will show that some problems have a middling success rate being tackled with either basinhopping or differential evolution.
I think the swapping is worth addressing, but I would classify it is a minor, not severe, bug.

I'm using macOS, it refused to open the posted zips, and besides it's safer to see the plain text gist.

**Comment by yonadav** [comment]
I can live with 'doesn't find the global minimum'.
However, if 2 codes, in this case with the swapping and without, both reach in some cases the global minimum and with the same number of iteration or nearly the same number, but one of then converges to the global minimum in more cases then it is probably better.
Anyhow, it is clearly a bug and you have fixed it.
The effect of the bug may somewhat resemble using a small mutation factor and thus more likely to converge to a local minimum. 
You don't really  need my example files. Use the Ackley function with popsize=7 (even 10 will demonstrate the difference), and the range of both variables is set to -500, 500. Test it on the previous version and the one with the swapping using the same seed for comparison purposes.
I could use tar.gz or any other method, next time will use gist.
